contrib/babelfishpg_tsql/src/pl_handler.c

```diff
@@ -3851,16 +3871,215 @@ bbf_ProcessUtility(PlannedStmt *pstmt,
 				}
 				break;
 			}
+		case T_GrantStmt:
+			{
+				GrantStmt *grant = (GrantStmt *) parsetree;
+				char	   *dbname = get_cur_db_name();
+				const char *current_user = GetUserNameFromId(GetUserId(), false);
+				/* Ignore when GRANT statement has no specific named object. */
+				if (sql_dialect != SQL_DIALECT_TSQL || grant->targtype != ACL_TARGET_OBJECT)
+					break;
+				Assert(list_length(grant->objects) == 1);
+				if (grant->objtype == OBJECT_SCHEMA)
+						break;
+				else if (grant->objtype == OBJECT_TABLE && strcmp("(CREATE LOGICAL DATABASE )", queryString) != 0)
+				{
+					/*
+					 * Ignore GRANT statements that are executed implicitly as a part of
+					 * CREATE database statements. Refer: create_bbf_db_internal().
+					 * These GRANT statement are just executed at the end, without checking any
+					 * schema permission or adding catalog entry.
+					 */
+					RangeVar   *rv = (RangeVar *) linitial(grant->objects);
+					const char *logical_schema = NULL;
+					char	   *obj = rv->relname;
+					bool exec_pg_command = false;
+					ListCell   *lc;
+					ListCell	*lc1;
+					if (rv->schemaname != NULL)
+						logical_schema = get_logical_schema_name(rv->schemaname, true);
+					else
+						logical_schema = get_authid_user_ext_schema_name(dbname, current_user);
+
+					/* If ALL PRIVILEGES is granted/revoked. */
+					if (list_length(grant->privileges) == 0)
+					{
+						if (grant->is_grant)
+						{
+							foreach(lc, grant->grantees)
+							{
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								add_or_update_object_in_bbf_schema(logical_schema, obj, ALL_PERMISSIONS_ON_RELATION, rol_spec->rolename, OBJ_RELATION, true, NULL);
+							}
+						}
+						else
+						{
+							foreach(lc, grant->grantees)
+							{
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								/*
+								 * 1. If permission on schema exists, don't revoke any permission from the object.
+								 * 2. If permission on object exists, update the privilege in the catalog and revoke permission.
+								 */
+								update_privileges_of_object(logical_schema, obj, ALL_PERMISSIONS_ON_RELATION, rol_spec->rolename, OBJ_RELATION, false);
+								if (privilege_exists_in_bbf_schema_permissions(logical_schema, PERMISSIONS_FOR_ALL_OBJECTS_IN_SCHEMA, rol_spec->rolename, false))
+									return;
+							}
+						}
+						exec_pg_command = true;
+					}
+					foreach(lc1, grant->privileges)
+					{
+						AccessPriv *ap = (AccessPriv *) lfirst(lc1);
+						AclMode privilege = string_to_privilege(ap->priv_name);
+						if (grant->is_grant)
+						{
+							exec_pg_command = true;
+							foreach(lc, grant->grantees)
+							{
+								RoleSpec	   *rol_spec = (RoleSpec *) lfirst(lc);
+								/* Don't add/update an entry, if the permission is granted on column list.*/
+								if (ap->cols == NULL)
```

Comment:
This is a loop-invariant. Skip the loop if ap->cols != NULL. I thought you'll take care of all such occurrences.